### PR TITLE
refactor: 重构生息演算

### PR DESCRIPTION
**Pull request by Alan-Charred**
Todo(s):
- [ ] Global Adaption
- [x] Gui Adapation
- [ ] Update Document

### CODE REVIEW

**Comment by Constrat** [comment]
Tales?

**Comment by Alan-Charred** [comment]
> Tales?

Tales within the sand

**Comment by Constrat** [comment]
Oh

**Comment by Constrat** [comment]
What are all those new fields? Isn't this just refactoring?

**Comment by Alan-Charred** [comment]
@MaaAssistantArknights/global-support-team 

Dear MAA global support team,

In this PR I rewrite the task chain for `Reclamation Algorithm: Tales Within The Sand`. Could you please help with the global adaptation? I have put placeholders at the end of every tasks.json file as depicted below.

image

I hope the text and task names have explained themselves.
Please let me know if I can do more to clarify the scenes where the corresponding text occurs.

Many thanks,

**Comment by Constrat** [comment]
Also I really don't think ocrReplace is the correct choice? Why not use the text field where possible?

**Comment by Alan-Charred** [comment]
> Also I really don't think ocrReplace is the correct choice? Why not use the text field where possible?

That's a decent argument. Let me replace the placeholders with text fileds.

Cheers,

**Comment by Constrat** [comment]
Also, this doesn't seem just refactoring, I'm noticing on the cn a lot of text changes that don't follow the original. Meaning that the global adaptation needs to redone (hopefully not from zero) is there a need for all this hassle?

**Comment by HX3N** [comment]
Recently, there's been so much refactoring in Reclamation that I have no idea what the tasks are doing anymore 🤣

**Comment by Alan-Charred** [comment]
> Super reclamation algorithm text guide



> Also, this doesn't seem just refactoring, I'm noticing on the CN a lot of text changes that don't follow the original. I can't find any old Chinese texts in the new ones you pushed: meaning that the global adaptation needs to redone (hopefully not from zero) is there a need for all this hassle?
> 
> Refactoring is good when it's easy to be done and pretty automatic, meaning just changing task names and order, but not changing the inside content. Especially on the global side.
> 
> * Task names are different (not ideal, but still decent)
> * Task content is different (not good)
> * Task content has no connection to older content (even worse)
> 
> @HX3N [Super reclamation algorithm text guide](https://github.com/MaaAssistantArknights/MaaAssistantArknights/issues/9992#issuecomment-2267951824) is now useless :(

Actually, you are right.
I rewrite the entire thing without referring to the old one, to ensure the maximum possibility of reusing these tasks in the future.
This is a first step of, making RA functionality as useful as the IS ones… though I don't have a good argument how necessary it is.
This is more like… rewrite stuff in a more elegant way (in my opinion) just because we could and we should keep the elegant stuff.
Hmm, I will try retrieve some of the texts from the old tasks to reduce extra efforts... Actually I can do it... by visiting the retrieved game data?

**Comment by Alan-Charred** [comment]
> Recently, there's been so much refactoring in Reclamation that I have no idea what the tasks are doing anymore 🤣
> 
> > In this PR I rewrite the task chain for Reclamation Algorithm: Tales Within The Sand. Could you please help with the global adaptation? I have put placeholders at the end of every tasks.json file as depicted below.
> 
> Btw, I just need to add the text or ocrReplace for that task, right?

Ah,,, actually... I changed my mind. I can handle this global adaptation myself. I will appreciate if you let me know by any means the issues about RA after this PR.

Cheers,

**Comment by Constrat** [comment]
Personally adapting from CN is always a hassle, because the easiest way would be to also have a levelled CN account and follow the ROI; but I stopped managing a CN account I had because it was a pain in the ass, so I have to rely on random screenshots or HX3N beautiful work XD. Because unfortunately when new stuff is implemented in CN, it's just implemented, where the adaptation docs "are missing" (they never existed in the first place).

A very important idea and "must" would be to obligated posting screens with ROI of all the tasks that require OCR and text to simplify the globalization process, but it's a lot to ask for. That lot to ask for then is pushed on us global maintainers.

I have no issue with doing it; I'll do it if it's absolutely necessary. However, if it isn't, it'll feel like a waste of time.

**Comment by Alan-Charred** [comment]
> Personally, adapting from CN is always a hassle, because the easiest way would be to also have a levelled CN account and follow the ROI; but I stopped managing a CN account I had because it was a pain in the ass, so I have to rely on random screenshots or HX3N beautiful work XD. Because unfortunately when new stuff is implemented in CN, it's just implemented, where the adaptation docs "are missing" (they never existed in the first place).
> 
> A very important idea and "must" would be to obligated posting screens with ROI of all the tasks that require OCR and text to simplify the globalization process, but it's a lot to ask for. That lot to ask for then is pushed on us global maintainers.
> 
> I have no issue with doing it; I'll do it if it's absolutely necessary. However, if it isn't, it'll feel like a waste of time.

On the other hand, I have started keeping managing an EN account, not only for global adaptation, but also because I want to keep variable names making sense — in that I use terms occurs in EN client.
But I can only do that for EN since I don't understand other languages.

Anyway my apologise for the inconvenience.

**Comment by Alan-Charred** [comment]
@MaaAssistantArknights/global-support-team 

Dear MAA global support team,

I think I have filled in most of fields except one for JP and KR.
Could you please find the text depicted below and fill it in the text field of task `Tales@Reclamation@CancelEnterBuildMode`?
Just the text on the bottom right button.

IMG_6409

And again, if by any chance this PR is merged, please let me know when you find anything wrong with this new task chain. I hope I have make it much more organised and as robust as possible.

Sincere Thanks,

**Comment by SherkeyXD** [comment]
GUI 有什么要改的

**Comment by Alan-Charred** [comment]
> GUI 有什么要改的

参数名字和类型需要改。我来看看吧，届时您帮我review一下，我正好去切个系统看看为什么windows CI有点问题。怀疑是我手动改vcxproj file 写错了什么。感谢啦～

File: resource/tasks.json [hunk 1]
```diff
@@ -2209,11 +2209,33 @@
...
     },
+    "RecogniseLoadingIcon": {
+        "baseTask": "LoadingIcon",
+        "template": "LoadingIcon.png",
+        "next": ["AfterLoadingIcon"]
+    },
+    "AfterLoadingIcon": {
+        "algorithm": "JustReturn",
+        "next": ["RecogniseLoadingIcon", "#back"]
+    },
```
Comment (Daydreamer114) [inline resource/tasks.json:2233]: 可以写成"next": ["#self", "#back"]吗

File: src/MaaCore/MaaCore.vcxproj.filters [hunk 2]
```diff
@@ -465,9 +465,6 @@
     <ClInclude Include="Task\Roguelike\Map\RoguelikeRoutingTask.h">
       <Filter>Source\Task\Roguelike</Filter>
     </ClInclude>
-    <ClInclude Include="Task\Roguelike\RoguelikeConfig.h">
-      <Filter>Source\Task\Roguelike</Filter>
-    </ClInclude>
```
Comment (status102) [inline src/MaaCore/MaaCore.vcxproj.filters:470]: ？

File: src/MaaCore/Task/Reclamation/ReclamationConfig.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    // ———————— ReclamationStrategy::ProsperityInSave 专用参数 ————————————————————————————
+    if (m_strategy == ReclamationStrategy::ProsperityInSave) {
+        m_tool_to_craft = params.get("tool_to_craft", "荧光棒");
+    }
+
+    return true;
+}
+
+void ReclamationConfig::set_theme(const std::string& theme)
```
Comment (status102) [inline src/MaaCore/Task/Reclamation/ReclamationConfig.cpp:37]: 没看到使用

File: src/MaaCore/Task/Reclamation/ReclamationConfig.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    static constexpr std::string_view Tales = "Tales"; // Tales Within The Sand
+};
+
+enum class ReclamationStrategy // 对应 Roguelike Mode
+{
+    ProsperityNoSave = 0,      // 0 - 无存档刷繁荣点数
+    ProsperityInSave = 1,      // 1 - 有存档刷繁荣点数
+};
+
+enum class ReclamationMode // 对应 Roguelike Difficulty
```
Comment (status102) [inline src/MaaCore/Task/Reclamation/ReclamationConfig.h:19]: 那就叫Difficulty好了，参照肉鸽

File: resource/tasks.json (hunk 1 above)
Comment (status102) [inline resource/tasks.json:2233]: 那为啥不直接用LoadingIcon

File: src/MaaCore/Task/Reclamation/ReclamationConfig.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+namespace asst
+{
+class ReclamationTheme
+{
+public:
+    static constexpr std::string_view Fire = "Fire";   // Fire Within The Sand
+    static constexpr std::string_view Tales = "Tales"; // Tales Within The Sand
+};
+
+enum class ReclamationStrategy // 对应 Roguelike Mode
```
Comment (status102) [inline src/MaaCore/Task/Reclamation/ReclamationConfig.h:13]:
```suggestion
enum class ReclamationMode // 对应 Roguelike Mode
```

File: src/MaaCore/Task/Reclamation/ReclamationConfig.h (hunk 4 above)
Comment (status102) [inline src/MaaCore/Task/Reclamation/ReclamationConfig.h:19]:
```suggestion
enum class ReclamationDifficulty // 对应 Roguelike Difficulty
```

**Comment by Lancarus** [comment]
> > Tales?
> 
> Tales within the sand

why not ra2

**Comment by Alan-Charred** [comment]
> > > Tales?
> > 
> > 
> > Tales within the sand
> 
> why not ra2

Because *I think* theme tags is better than pure numbering.

File: resource/tasks.json (hunk 1 above)
Comment (Alan-Charred) [reply resource/tasks.json:2233]: 现在看来，唯一的 argument 是我不喜欢 LoadingIcon.next 里面那个 #next, 这让我感觉对实际 task 行为不可控。如果没有那个 #next 的话，你是对的，完全可以
```json
"AfterLoadingIcon": {
    "algorithm": "JustReturn",
    "next": ["LoadingIcon", "#back"]
},
```

File: src/MaaCore/Task/Reclamation/ReclamationConfig.cpp (hunk 3 above)
Comment (Alan-Charred) [reply src/MaaCore/Task/Reclamation/ReclamationConfig.cpp:37]: 保留了肉鸽 Config 的风格，我觉得的确可以删。因为这些参数在 set_params 后基本都是只读的。
(肉鸽现在好像也不用 set_theme 了)。
先留着吧要不？

File: resource/tasks.json (hunk 1 above)
Comment (Alan-Charred) [reply resource/tasks.json:2233]: 好像不太一样。但是我其他一些任务，A.next 里包含 A 自身的，的确可以换成 #self 以追求更简洁。

File: src/MaaCore/Task/Reclamation/ReclamationConfig.h (hunk 5 above)
Comment (Alan-Charred) [reply src/MaaCore/Task/Reclamation/ReclamationConfig.h:13]: 老师您让我留着这个吧，和 EN 的 文字描述保持一致。我都自己改 GUI 了（戳手
反正这块儿以后我来负责糊（

Comment (status102) [reply src/MaaCore/Task/Reclamation/ReclamationConfig.h:13]: > 老师您让我留着这个吧，和 EN 的 文字描述保持一致。我都自己改 GUI 了（戳手 反正这块儿以后我来负责糊（

那约定传参是mode对应策略，difficulty对应难度

File: src/MaaCore/Task/Reclamation/ReclamationConfig.cpp (hunk 3 above)
Comment (status102) [reply src/MaaCore/Task/Reclamation/ReclamationConfig.cpp:37]: > 保留了肉鸽 Config 的风格，我觉得的确可以删。因为这些参数在 set_params 后基本都是只读的。 (肉鸽现在好像也不用 set_theme 了)。 先留着吧要不？

肉鸽是改了读参，现在也不用了，等会去删了

File: src/MaaCore/Task/Reclamation/ReclamationConfig.h (hunk 5 above)
Comment (status102) [reply src/MaaCore/Task/Reclamation/ReclamationConfig.h:13]: > 老师您让我留着这个吧，和 EN 的 文字描述保持一致。我都自己改 GUI 了（戳手 反正这块儿以后我来负责糊（

这个还是先改掉吧，特别是传参。后面改传参怪麻烦的

File: src/MaaCore/MaaCore.vcxproj.filters (hunk 2 above)
Comment (Alan-Charred) [reply src/MaaCore/MaaCore.vcxproj.filters:470]: 老师说出来您不信，这是一个不得不手动修改 vcxproj 文件的人随手删掉了重复的内容。里面现在 600+ 行还有一个这个
image

File: resource/tasks.json [hunk 6]
```diff
@@ -11563,5 +11571,649 @@
...
+    "Tales@Reclamation@LeaveCurrentZone": {
+        "doc": "模板任务: 在行动中点击左上角离开当前区块",
+        "roi": [41, 22, 78, 71],
+        "action": "ClickSelf"
+    },
+    "Tales@Reclamation@ProsperityNoSave@ConfirmToLeaveBaseCamp": {
+        "doc": "确认离开 <后舍: 众人会聚之地>",
+        "baseTask": "Tales@Reclamation@DialogConfirmYellow",
+        "template": "Tales@Reclamation@DialogConfirmYellow.png",
+        "next": ["Tales@Reclamation@ProsperityNoSave@ClickShrubberyGalore@AfterLoadingIcon@AfterLoadingText"]
```
[the remark this answers is not in the document]
Comment (Alan-Charred) [reply resource/tasks.json:11644]: 还是不太对，但是我有思路了。

**Comment by Constrat** [comment]
> > Because _I think_ theme tags is better than pure numbering.
> 
> This is not like the IS....I mean only one word “tales” sometimes may not remind others of “Tales within the sand”， not like themes in rouguelike.

Strongly agree with this one. Phantom, Mizuki, Sami, Sarkaz are all clear indicators (especially since after there's `@Roguelike`. Here there's `@Reclamation` but in english "Tales" literally means "little story" so it doesn't make sense.

Plus I don't think anyone here differentiaties RA1 and RA2 by their names (Hell I always forget one is Fire within the Sand or whatever and the other is Tales... I forget XD)

**Comment by Alan-Charred** [comment]
> > This is not like the IS....I mean only one word “tales” sometimes may not remind others of “Tales within the sand”， not like themes in rouguelike.
> 
> Strongly agree with this one. Phantom, Mizuki, Sami, Sarkaz are all clear indicators (especially since after there's `@Roguelike`. Here there's `@Reclamation` but in english "Tales" literally means "little story" so it doesn't make sense.
> 

1. As menthoned in https://github.com/MaaAssistantArknights/MaaAssistantArknights/pull/10215, the first tag I proposed is SandFire & SandTales. After the discussion on the order of terms, I choose to walk around by removing the term "sand".

2. We should have a consistent naming style for iterative play mode throughout the entire project.
If you guys also apply the naming rule of "PlayModeTag + Number" to integreted strategy, then I'm in.

3. So far these is only one playable season for RA. No tag makes sense if it is the only useable one. I myself do not know whether "Fire", "SandFire", "FireSand", "Tales", "SandTales", "TalesSand" are good until the next permanent season is released.

4. The available seasons starting from number 2 is also a source of confusion. Hmm, This reminds me of the Lakeview Cabin series... Alas, it should be more intuitive to number the first experimental season as 0.

5. As you said in earlier comments, this kind of refactoring/rewriting/reimplementing could be a waste of time. My original intention is to pave the way towards future functionalities and scalability. However, this is just like having a mapping between two singleton sets. People generally do not value this until the more elements comes, until the essential needs approach. As a student in theoretical research, I totally understand this.

In conclusion, I would like to close this PR. Feel free to revert my changes in https://github.com/MaaAssistantArknights/MaaAssistantArknights/pull/10215. Let's continue with `Reclamation@___` and `Reclamation2@___`, which I guess is more widely desired.

> Plus I don't think anyone here differentiaties RA1 and RA2 by their names (Hell I always forget one is Fire within the Sand or whatever and the other is Tales... I forget XD)

I do. I enjoy this play mode and have spent a lot of time in it. Others do not differentiate RA1 and RA2 perhaps because only one of them is currently available, and this play mode is not as popular as IS.

I wonder whether the same argument occurs when there is only one `RoguelikeTheme::Phantom`. And what have you done to `RoguelikeTheme::Ceobe`?


Regards,
